sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -0,0 +1,378 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.schemas;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.ImmutableList;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collector;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.coders.RowCoder;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * {@link Schema} describes the fields in {@link Row}.
+ *
+ */
+@Experimental
+@AutoValue
+public abstract class Schema implements Serializable {
+  // A mapping between field names an indices.
+  private BiMap<String, Integer> fieldIndices = HashBiMap.create();
+  public abstract List<Field> getFields();
+
+  @AutoValue.Builder
+  abstract static class Builder {
+    abstract Builder setFields(List<Field> fields);
+    abstract Schema build();
+  }
+
+  public static Schema of(List<Field> fields) {
+    return Schema.fromFields(fields);
+  }
+
+  public static Schema of(Field ... fields) {
+    return Schema.of(Arrays.asList(fields));
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (!(o instanceof Schema)) {
+      return false;
+    }
+    Schema other = (Schema) o;
+    return Objects.equals(fieldIndices, other.fieldIndices)
+        && Objects.equals(getFields(), other.getFields());
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(fieldIndices, getFields());
+  }
+
+  /**
+   * An enumerated list of supported types.
+   */
+  public enum TypeName {
+    BYTE,    // One-byte signed integer.
+    INT16,   // two-byte signed integer.
+    INT32,   // four-byte signed integer.
+    INT64,   // eight-byte signed integer.
+    DECIMAL,  // Decimal integer
+    FLOAT,
+    DOUBLE,
+    STRING,  // String.
+    DATETIME, // Date and time.
+    BOOLEAN,  // Boolean.
+    ARRAY,
+    ROW;    // The field is itself a nested row.
+
+    public static final List<TypeName> NUMERIC_TYPES = ImmutableList.of(
+        BYTE, INT16, INT32, INT64, DECIMAL, FLOAT, DOUBLE);
+    public static final List<TypeName> STRING_TYPES = ImmutableList.of(STRING);
+    public static final List<TypeName> DATE_TYPES = ImmutableList.of(DATETIME);
+    public static final List<TypeName> CONTAINER_TYPES = ImmutableList.of(ARRAY);
+    public static final List<TypeName> COMPOSITE_TYPES = ImmutableList.of(ROW);
+
+    public boolean isNumericType() {
+      return NUMERIC_TYPES.contains(this);
+    }
```

Comment:
Github is not feeling well, I may have posted this question in another thread. These methods are probably invoked a lot, should we change these to flags? E.g.:

```java
enum TypeName {
  BYTE(true, false),
  STRING(false, true);

  private boolean isNumeric;
  private boolean isString;

  TypeName(boolean isNumeric, boolean isString) { ... }
}
```